src/Functions/formatDateTime.cpp

```diff
@@ -497,15 +497,20 @@ class FunctionFormatDateTimeImpl : public IFunction
 
         size_t mysqlFractionalSecond(char * dest, Time /*source*/, UInt64 fractional_second, UInt32 scale, const DateLUTImpl & /*timezone*/)
         {
-            if (scale == 0)
-                scale = 6;
+            // if more than 6 digits are passed in, we truncate to 6 == buffer size
+            while (scale > 6)
+            {
+                --scale;
+                fractional_second /= 10;
+            }
 
-            for (Int64 i = scale, value = fractional_second; i > 0; --i)
+            // mysql format option %f always produces six digits. scale only defines which of them get filled
+            for (UInt32 i = scale, value = fractional_second; i > 0 && value > 0; --i)
```

Comment:
Should we fill all 6 digits then? If scale is less than 6 we might have some bits unitiailized, I think it makes sense to fill those places out also (with `0`).

Reply:
The buffer is initialized with 6 '0' characters (see line 1238 below). So, if we stop early here (i.e. value is zero), we keep the '0' in place. Note that this is part of the test-cases where I use scale < 6.